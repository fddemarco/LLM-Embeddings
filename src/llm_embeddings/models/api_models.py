import voyageai


class VoyageAiModel:
    def __init__(self, api_key) -> None:
        self.client = voyageai.Client(api_key=api_key)

    def embed(self, texts):
        if not texts:
            raise ValueError("Input cannot be empty list")

        if len(texts) > 128:
            raise ValueError(
                f"The batch size limit is 128. Your batch size is {len(texts)}"
            )

        return self.client.embed(
            texts, model="voyage-lite-01", input_type="document"
        ).embeddings
