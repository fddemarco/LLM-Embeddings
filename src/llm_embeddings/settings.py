import pathlib

from decouple import config

API_KEY = None
API_KEY_NAME = "API_KEY"
DATA_PATH = "DATA_PATH"


def get_key_name():
    return API_KEY or config(API_KEY_NAME)


def get_api_key():
    return config(get_key_name())


def get_data_path():
    return config("DATA_PATH", cast=pathlib.Path, default="kaggle/working/")
