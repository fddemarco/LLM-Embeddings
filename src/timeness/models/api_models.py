import openai
import tiktoken
import voyageai


class Model:
    def embed(self, sentences):
        self.assert_are_valid(sentences)
        sentences = self.preprocess(sentences)
        sentences = self.truncate(sentences)
        return self.safe_embed(sentences)

    def assert_are_valid(self, sentences):
        self.assert_not_empty(sentences)
        self.assert_allowed_batch_size(sentences)

    def assert_not_empty(self, sentences):
        if not sentences:
            raise ValueError("Input cannot be empty list")

    def assert_allowed_batch_size(self, sentences):
        if len(sentences) > 128:
            raise ValueError(
                f"The batch size limit is 128. Your batch size is {len(sentences)}"
            )

    def safe_embed(self, sentences):
        raise NotImplemented("Should be implemented in a subclass")

    def preprocess(self, sentences):
        return [sentence.replace("\n", " ") for sentence in sentences]

    def truncate(self, sentences):
        raise NotImplemented("Should be implemented in a subclass")


class VoyageAiModel(Model):
    def __init__(self, api_key) -> None:
        self.client = voyageai.Client(api_key=api_key)
        self.n_dimensions = 1024

    def safe_embed(self, sentences):
        response = self.get_response(sentences)
        return response.embeddings

    def get_response(self, sentences):
        return self.client.embed(
            sentences, model="voyage-lite-01", input_type="document", truncation=True
        )

    def truncate(self, sentences):
        return sentences


class Ada2Model(Model):
    def __init__(self, api_key):
        self.client = openai.OpenAI(api_key=api_key)
        self.max_input_tokens = 8192
        self.n_dimensions = 1536
        self.embedding_encoding = "cl100k_base"

    def safe_embed(self, sentences):
        response = self.get_response(sentences)
        return [data.embedding for data in response.data]

    def get_tokens_count(self, sentences):
        response = self.get_response(sentences)
        return [usage for _, usage in response.usage]

    def get_response(self, sentences):
        return self.client.embeddings.create(
            input=sentences, model="text-embedding-ada-002"
        )

    def truncate(self, sentences):
        encoder = tiktoken.get_encoding(self.embedding_encoding)
        return [
            encoder.encode(sentence)[: self.max_input_tokens] for sentence in sentences
        ]
