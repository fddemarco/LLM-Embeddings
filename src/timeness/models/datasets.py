import polars as pl


class Schema:
    def __init__(self, text_key, group_key):
        self.text_key = text_key
        self.group_key = group_key


class Dataset:
    def __init__(self, data, schema):
        self.data = data
        self.schema = schema

    def embed_text(self, model):
        return (
            self.data.with_columns(
                pl.col(self.schema.text_key)
                .map_elements(lambda text: model.embed([text])[0])
                .alias("temp")
            )
            .with_columns(
                pl.col("temp").list.to_struct(
                    fields=[str(i) for i in range(0, model.n_dimensions)]
                )
            )
            .unnest("temp")
        )
