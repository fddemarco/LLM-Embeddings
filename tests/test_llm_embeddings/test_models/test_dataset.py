import polars as pl
import pytest

from llm_embeddings import settings
from llm_embeddings.models import api_models
from llm_embeddings.models import datasets


@pytest.fixture(name="schema")
def fixture_schema():
    return datasets.Schema("text", "group")


@pytest.fixture(name="data")
def fixture_data(schema):
    return pl.DataFrame(
        {schema.text_key: ["Sample text 1", "Sample text 2"], schema.group_key: [0, 1]}
    )


@pytest.fixture(name="dataset")
def fixture_dataset(data, schema):
    return datasets.Dataset(data, schema)


@pytest.fixture(name="api_key")
def fixture_api_key():
    return settings.get_config(settings.VOYAGEAI_API_KEY)


@pytest.fixture(name="model")
def fixture_model(api_key):
    return api_models.VoyageAiModel(api_key)


@pytest.fixture(name="openai_api_key")
def fixture_api_key():
    return settings.get_config(settings.OPENAI_API_KEY)


@pytest.fixture(name="ada_model")
def fixture_model(openai_api_key):
    return api_models.Ada2Model(openai_api_key)


@pytest.mark.slow
def test_01(snapshot, dataset, model):
    assert snapshot == dataset.embed_text(model)


def test_02(snapshot, dataset, ada_model):
    assert snapshot == dataset.embed_text(ada_model)
