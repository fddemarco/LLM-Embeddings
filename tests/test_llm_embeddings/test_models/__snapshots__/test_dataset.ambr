# serializer version: 1
# name: test_01
  shape: (2, 1_026)
  ┌───────────────┬───────┬──────────┬───────────┬───┬───────────┬───────────┬──────────┬───────────┐
  │ text          ┆ group ┆ 0        ┆ 1         ┆ … ┆ 1020      ┆ 1021      ┆ 1022     ┆ 1023      │
  │ ---           ┆ ---   ┆ ---      ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---      ┆ ---       │
  │ str           ┆ i64   ┆ f64      ┆ f64       ┆   ┆ f64       ┆ f64       ┆ f64      ┆ f64       │
  ╞═══════════════╪═══════╪══════════╪═══════════╪═══╪═══════════╪═══════════╪══════════╪═══════════╡
  │ Sample text 1 ┆ 0     ┆ 0.036505 ┆ -0.026271 ┆ … ┆ -0.004697 ┆ -0.066844 ┆ 0.018421 ┆ -0.052074 │
  │ Sample text 2 ┆ 1     ┆ 0.033346 ┆ -0.030889 ┆ … ┆ -0.001532 ┆ -0.067521 ┆ 0.016831 ┆ -0.050024 │
  └───────────────┴───────┴──────────┴───────────┴───┴───────────┴───────────┴──────────┴───────────┘
# ---
# name: test_02
  shape: (2, 1_538)
  ┌──────────────┬───────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
  │ text         ┆ group ┆ 0         ┆ 1         ┆ … ┆ 1532      ┆ 1533      ┆ 1534      ┆ 1535      │
  │ ---          ┆ ---   ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
  │ str          ┆ i64   ┆ f64       ┆ f64       ┆   ┆ f64       ┆ f64       ┆ f64       ┆ f64       │
  ╞══════════════╪═══════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
  │ Sample text  ┆ 0     ┆ -0.001273 ┆ -0.00472  ┆ … ┆ -0.003565 ┆ -0.009776 ┆ -0.008117 ┆ -0.015726 │
  │ 1            ┆       ┆           ┆           ┆   ┆           ┆           ┆           ┆           │
  │ Sample text  ┆ 1     ┆ -0.013762 ┆ -0.009575 ┆ … ┆ -0.007207 ┆ -0.016137 ┆ -0.022951 ┆ -0.010946 │
  │ 2            ┆       ┆           ┆           ┆   ┆           ┆           ┆           ┆           │
  └──────────────┴───────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴───────────┘
# ---
